File: fastapi_server/fastapi_server/bootstrap/registry/loader_logger.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from fastapi_server.bootstrap.contract import LoaderReport, report_error
from fastapi_server.bootstrap.registry.ctx import Logger


@dataclass
class LoaderLogger:
    addon_name: str
    logger: Logger
    report: LoaderReport

    @property
    def _tag(self) -> str:
        return f"[{self.addon_name}]"

    def scan_dir(self, directory: str, matched: int, ignored: int) -> None:
        self.logger.debug(f"{self._tag} scan {directory}: matched={matched} ignored={ignored}")

    def ignored(self, path: str) -> None:
        self.logger.debug(f"{self._tag} ignored {path}")

    def loaded(self, path: str) -> None:
        self.logger.debug(f"{self._tag} loaded {path}")

    def registered(self, path: str, kind: Optional[str] = None) -> None:
        suffix = f" ({kind})" if kind else ""
        self.logger.debug(f"{self._tag} registered {path}{suffix}")

    def skipped(self, path: str, reason: str) -> None:
        self.logger.info(f"{self._tag} skipped {path}: {reason}")

    def failed(self, step: str, target: Optional[str], err: object) -> None:
        msg = str(err) if not isinstance(err, BaseException) else (str(err) or err.__class__.__name__)
        target_part = f" {target}" if target else ""
        # error() so the addon's swallowed failures land on stderr — without
        # this, ImportError-level breakage looks like normal stdout output.
        self.logger.error(f"{self._tag} FAIL [{step}]{target_part}: {msg}")
        report_error(self.report, step, err, target)


def create_loader_logger(
    addon_name: str,
    logger: Logger,
    report: LoaderReport,
) -> LoaderLogger:
    return LoaderLogger(addon_name=addon_name, logger=logger, report=report)

File: fastify_server/src/registry/loader_logger.ts
import type { LoaderReport } from "../contract/index.js";
import { reportError } from "../contract/index.js";
import type { Logger } from "./ctx.js";

export interface LoaderLogger {
  scanDir(dir: string, matched: number, ignored: number): void;
  ignored(path: string): void;
  loaded(path: string): void;
  registered(path: string, kind?: string): void;
  skipped(path: string, reason: string): void;
  failed(step: string, target: string | undefined, err: unknown): void;
}

export function createLoaderLogger(
  addonName: string,
  logger: Logger,
  report: LoaderReport,
): LoaderLogger {
  const tag = `[${addonName}]`;
  return {
    scanDir(dir, matched, ignored) {
      logger.debug(`${tag} scan ${dir}: matched=${matched} ignored=${ignored}`);
    },
    ignored(path) {
      logger.debug(`${tag} ignored ${path}`);
    },
    loaded(path) {
      logger.debug(`${tag} loaded ${path}`);
    },
    registered(path, kind) {
      logger.debug(`${tag} registered ${path}${kind ? ` (${kind})` : ""}`);
    },
    skipped(path, reason) {
      logger.info(`${tag} skipped ${path}: ${reason}`);
    },
    failed(step, target, err) {
      const msg = err instanceof Error ? err.message : String(err);
      // error() so swallowed addon failures land on stderr (fastify.error.log)
      // instead of looking like ordinary stdout. Pass the raw err so the
      // consoleLogger surfaces stack/cause on a follow-up line.
      logger.error(`${tag} FAIL [${step}]${target ? ` ${target}` : ""}: ${msg}`, err);
      reportError(report, step, err, target);
    },
  };
}
